src/GraphQL/API/Enum.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DefaultSignatures #-}
+
+module GraphQL.API.Enum
+  ( GraphQLEnum(..)
+  ) where
+
+import Protolude hiding (Enum, U1, TypeError)
+import GraphQL.Internal.AST (Name, nameFromSymbol, NameError)
+import qualified GraphQL.Value as GValue
+import GHC.Generics (M1(..), Rep, Meta(..), C1, U1(..), D, (:+:)(..))
+import GHC.TypeLits (KnownSymbol, TypeError, ErrorMessage(..))
+
+
+class GenricEnumValues (f :: Type -> Type) where
+  genericEnumValues :: [Either NameError Name]
+  genericEnumFromValue :: GValue.Value -> Either Text (f p)
+  genericEnumToValue :: f p -> GValue.Value
+
+instance forall conName m p f nt.
+  ( KnownSymbol conName
+  , KnownSymbol m
+  , KnownSymbol p
+  , GenricEnumValues f
+  ) => GenricEnumValues (M1 D ('MetaData conName m p nt) f) where
+  genericEnumValues = genericEnumValues @f
+  genericEnumFromValue v@(GValue.ValueEnum _) = M1 <$> genericEnumFromValue v
+  genericEnumFromValue x = Left ("Not an enum: " <> show x)
+  genericEnumToValue (M1 gv) = genericEnumToValue gv
+
+instance forall conName f p b.
+  ( KnownSymbol conName
+  , GenricEnumValues f
+  ) => GenricEnumValues (C1 ('MetaCons conName p b) U1 :+: f) where
+  genericEnumValues = let name = nameFromSymbol @conName in name:(genericEnumValues @f)
+  genericEnumFromValue v@(GValue.ValueEnum vname) =
+    case nameFromSymbol @conName of
+      Right name -> if name == vname
+                    then L1 <$> Right (M1 U1)
+                    else R1 <$> genericEnumFromValue v
+      Left x -> Left ("Not a valid enum name: " <> show x)
+  genericEnumFromValue _ = panic "This case should have been caught at top-level. Please file a bug."
+  genericEnumToValue (L1 _) =
+    let Right name = nameFromSymbol @conName
+    in GValue.ValueEnum name
+  genericEnumToValue (R1 gv) = genericEnumToValue gv
+
+instance forall conName p b. (KnownSymbol conName) => GenricEnumValues (C1 ('MetaCons conName p b) U1) where
+  genericEnumValues = let name = nameFromSymbol @conName in [name]
+  genericEnumFromValue (GValue.ValueEnum vname) =
+    case nameFromSymbol @conName of
+      Right name -> if name == vname
+                    then Right (M1 U1)
+                    else Left ("Not a valid enum name: " <> show vname)
+      Left x -> Left ("Not a valid enum name: " <> show x)
```

Comment:
done

I'm not sure I want to lose the information about the error (e.g. the name that didn't work), so not sure about the Maybe..